### ISSUE TITLE
Custom partitioning of historic data using table decorators

### ISSUE THREAD

**Issue by kurtmaile** [samelamin/spark-bigquery#33]
Hi Sam,

As a one off, I am migrating 8 years worth of historic data, ingesting flat files and transforming to the new model, with the idea of writting this data (statically - not streamed) to a number of partitioned table in BQ.  Ive been chatting to some of the BQ guys who suggested table decorators to enable setting a custom partition and thus get the performance gain in queries in BQ (note: this is actually needed in streaming to so that true event time is used and not ingest time....but this issue is for a static load)

Using the APIs, I need to:

1. Create a new paritioned table 'order_totals_all_currencies' - I was hoping the APIs can create if it doesnt exist like normal and autodetect / generate the BQ schema 9very cool feature BTW!), but also recognising the table decorator usage.  e.g "order_totals_all_currencies$20170717"

2. Then write the data, however use **table decorators** to guide the data to the correct event time partition, as mentioned its historical and needs to be saved using a partiton date set in my spark job to be when the order was placed e.g 20150301?  So I would be iterating, doing multiple 'saveAsBigQueryTable' per day with a different table decorator.

e.g 
```
df20160101.saveAsBigQueryTable("order_totals_all_currencies$20160101",true,0)
df20160102.saveAsBigQueryTable("order_totals_all_currencies$20160102",true,0)
df20160103.saveAsBigQueryTable("order_totals_all_currencies$20160103",true,0)
```

Is this possible / what is the best way to do it?  Creating the table and schema by hand first would take some time (some large schemas), was hoping the library could handle this scenario and the initial create that has decoration?  Is there a way I am missing perhaps?

I was trying below as a test

```
val bqOrderTotalsAllCurrenciesTable = "order_totals_all_currencies$20170717"

orderTotalsAllCurrenciesHistory20170717DF.saveAsBigQueryTable(s"$bqProjectId:$bqSdvPreAggsDataset.$bqOrderTotalsAllCurrenciesTable",true, 0)
```

but got this error from google.

'"Invalid table ID \"order_totals_all_currencies$20170717\". Table IDs must be alphanumeric (plus underscores) and must be at most 1024 characters long. Also, Table decorators cannot be used.",

thought it might create the table (just the first time of course) AND then load the partitoned data subsequent using the date decorator for the gorups of orders.

I need the saveAsTable to be aware of the decorator is most key.

Id actually want to be able to do this in streaming too to enable true event time on the streams where we get data arriving late.

Thoughts? 

Cheers

**Comment by samelamin** [comment]
YEs we can use table decorators using the api but this functionality is not part of the package at the moment. Sounds like what you are intending to do is a one off so I suggest using dataflow to write to the tables. no need to use spark at all if it is a one off 


Table decorators would be a good feature to work on so I am marking this issue as a feature request

**Comment by kurtmaile** [comment]
Cool thanks would be super cool - I dont know dataflow, my spark job is pretty long, and acutally reuses some of my library to do computations (like currency converting for example) - porting to a different API (Beam) would take a mear mortal like me ages - as much as I would love to get dirty with DF and be able to say I could pick it up like that and move to java (after Scala - yuk!) hehe.  Wish I could.

Ill just have to make do with non partitioned, thanks for replying 

p.s - still thinking about the offer, its possible if I could span out the payments longer perhaps?

**Comment by samelamin** [comment]
perhaps do all your transformations and save the result to partitioned files locally? then write a simple script to upload the tables to the correct tables using the bq command line? 


Yeah sure just drop me an email and we can work out the details

**Comment by samelamin** [comment]
this has been fixed with version 0.2.1. @kurtmaile

**Comment by kurtmaile** [comment]
Hey great thanks - I'm away this weekend will test on Monday! Exciting...
